### PR TITLE
Shikkic/short shas

### PR DESCRIPTION
**Pull request by Shikkic**
**Not ready to be merged, still need to fix all the tests.**

Just wanted to get a PR up early, will be focusing on tests and clean up now.

### CODE REVIEW

File: common/github/fetch_full_sha.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// FetchFullSHAFromPartialSHA is responsible for fetching a full commit SHA from a short SHA
+func FetchFullSHAFromPartialSHA(author, repo, shortSHA string) (string, error) {
+	client := &http.Client{}
+	archiveURL := fmt.Sprintf(
+		"https://github.com/%s/%s/archive/%s.zip",
+		author,
+		repo,
+		shortSHA,
+	)
+	req, err := http.NewRequest(
```
Comment (skeswa) [inline common/github/fetch_full_sha.go]: A comment explaining why this works would be nice.

File: common/github/fetch_full_sha.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	if err != nil || resp.StatusCode == 404 {
+		return "", err
+	}
+
+	eTagHeader := resp.Header.Get(etagHeader)
+	if len(eTagHeader) < shaLength {
+		return "", errors.New("Unable to retrieve full commit SHA, Etag header was incomplete or empty.")
+	}
+
+	fullSHA := eTagHeader[1 : len(eTagHeader)-1]
```
Comment (skeswa) [inline common/github/fetch_full_sha.go]: Comment explaining why please.

File: router/interfaces.go [hunk 3]
```diff
@@ -15,6 +15,9 @@ import (
 // refsDownloader is responsible for downloading the git refs for a package.
 type refsDownloader func(author, repo string) (common.Refs, error)
 
+// fullSHAFetcher is responsible for fetching a full commit SHA from a short SHA
```
Comment (skeswa) [inline router/interfaces.go]: Missing period.

File: router/package_request.go [hunk 4]
```diff
@@ -57,47 +58,64 @@ func newPackageRequest(args newPackageRequestArgs) (*packageRequest, error) {
...
 
-			// Find the best candidate.
-			bestCandidate := refs.Candidates.Best(parts.semverSelector)
-			if bestCandidate == nil {
-				return nil, NewNoSuchPackageVersionError(
-					parts.author,
-					parts.repo,
-					parts.semverSelector.String())
-			}
+	// Get and process all of the refs for this package.
```
Comment (skeswa) [inline router/package_request.go]: This is only necessary if there is a semver selector. Move it into the semver if block.

File: router/package_request.go [hunk 5]
```diff
@@ -57,47 +58,64 @@ func newPackageRequest(args newPackageRequestArgs) (*packageRequest, error) {
...
-			// Re-serialize the refs data with said candidate.
-			matchedSHA = bestCandidate.GitRefHash
-			matchedSHALabel = bestCandidate.String()
-		} else if parts.hasSHASelector() {
-			// Re-serialize the refs data with the sha.
-			matchedSHA = parts.shaSelector
-			// TODO(skeswa): investigate validating the ref to see if it actually
-			// exists.
+	// Set the default matched sha in case there is no semver selector.
+	matchedSHA = refs.MasterRefHash
```
Comment (skeswa) [inline router/package_request.go]: We decided that defaulting to master is wrong if there is a semver selector that fails to find a candidate. Amend this so this only happens if there is no semver selector **and** there is no sha selector.

File: router/package_request_parts.go [hunk 6]
```diff
@@ -13,8 +13,10 @@ import (
 const (
 	at                          = '@'
 	dot                         = '.'
+	hyphen                      = '-'
```
Comment (skeswa) [inline router/package_request_parts.go]: Bulge. BULGE.

File: router/package_request_parts.go [hunk 7]
```diff
@@ -13,8 +13,10 @@ import (
...
 	slash                       = '/'
 	shaLength                   = 40
+	shortSHALength              = 6
```
Comment (skeswa) [inline router/package_request_parts.go]: minShortSHALength.

Comment (skeswa) [inline router/package_request_parts.go]: Consider dropping it to 5.

Comment (skeswa) [inline router/package_request_parts.go]: Cooler sounding number.

File: router/package_request.go (hunk 4 above)
Comment (Shikkic) [inline router/package_request.go]: I know we talked about this but for historical documentation we need this done to retrieve master branch too.

File: router/package_request.go (hunk 5 above)
Comment (Shikkic) [inline router/package_request.go]: Will fix this up as per how we spoke earlier

File: router/package_request_parts.go (hunk 6 above)
Comment (Shikkic) [inline router/package_request_parts.go]: (´⊙o⊙`；)

File: router/package_request_parts.go (hunk 7 above)
Comment (Shikkic) [inline router/package_request_parts.go]: I tried 5 but it was more consistent with 6 characters. Can try again though. Also will change to `minShortSHALength.`

**Comment by Shikkic** [comment]
- [x] Added tests for `FetchFullSHAFromPartialSHA` and general fixes.

Things to do:
- [ ] Need to clean up the order of newPackageRequest
- [x] Fix up any remaining newPackageRequest tests

File: router/package_request_parts.go [hunk 8]
```diff
@@ -253,7 +253,7 @@ func isFullSHASelector(selector string) bool {
 
 func isShortSHASelector(selector string) bool {
 	// If it isn't a 6 character short SHA return false, it might be semvar
-	return len(selector) == shortSHALength && strings.IndexByte(selector, dot) == -1 && strings.IndexByte(selector, hyphen) == -1
+	return len(selector) == minShortSHALength && strings.IndexByte(selector, dot) == -1 && strings.IndexByte(selector, hyphen) == -1
```
Comment (skeswa) [inline router/package_request_parts.go]: This should be an inequality. Should be greater or equal to min short sha length and less than full sha length

**Pull request merged**